Response function "bypass"

buckle2000 opened the pull request:
improve document

kikito commented:
Hi, thanks for this. A couple notes:

* Please replace "... BEFORE any collision happens" by " ... *before* any collision happens" - (use asterisks, and not uppercase, for emphasis).
* Please remove the "wrong" part about "Calling other response functions". Leave just the "good" part.

Thanks again!

buckle2000 commented:
Done

buckle2000 commented:
You may want to add new test cases.

buckle2000 commented:
Wait...
This response is not robust.

buckle2000 commented:
Now it's better.
@kikito your code is really impeccable.

kikito commented:
(I have not forgotten about this. I am just a bit busy ATM. I might be able to review this on tuesday next week. Thanks a lot!)

kikito commented:
Hi, I have reviewed this code and I have several questions:

* I understand what `rect_isInShadow` does (make the box a "cross" by extruding its 4 faces out infinitely. The function returns whether a given point falls inside that cross). I don't understand *why* it's used on that if, which gives a regular slide, the same if the collision already provokes some movement. Is it really needed? If it is, then it should be declared next to the rest of the `rect_*` functions, not on top of `bypass`.
* I am also having difficulties understanding how the usage of `rect_getNearestCorner` works. First it gets the "other" nearest corner to the center of the item where the touch happens. Then it gets the corner which is nearest to that first corner. And then it uses those two corners to make a dx and dy. But then it changes dx and dy depending on the collision normal.
* The modification of dx and dy is also a bit confusing.
* I think the algorithm would be a bit more understandable if the variables had longer names (short names are ok when they are few and used in 2 or 3 lines. When you have a bunch of them in a 20+ algorithm, it's better to use longer variable names. I realize that I can have done that mistake myself.
* Are you planning to add any specs, or are those left as an exercise for me? :)

Regards!

buckle2000 commented:
**Is `rect_isInShadow` needed?**
Yes. It is used to determine whether the moving object should bypass other object. I have moved it to the correct place.

---
**What's the usage of `rect_getNearestCorner`?**
If I want to let `item` bypass `other`, I would like to first *align the two corners you said before* and then move `item` with velocity before touch.

---
**The modification of dx and dy is also a bit confusing**
Well...

---
`longer names`
Updated!
Comments added!

---
**specs**
I never use specs before and I have no idea how to add specs for this.
Just forget this.

---
I also fixed a bug during code revision.

Thanks for you feedback.

buckle2000 commented:
@kikito It works this time.

kikito commented:
(@buckle2000: again, I have not forgotten about this. This is something which deserves some time to be analyzed. I would have to test it manually a bit to see how it behaves (by the way, if you have a working example that you can share, that would speed things up for me). And I would like to write tests for this before I merge it with master. So, please be patient. With a litte bit of luck I might be able to scratch some hours during the weekend.)

buckle2000 commented:
The example:
https://github.com/buckle2000/bump.lua/tree/simpledemo

kikito commented:
Hi, thanks for the working example! I have toyed with it for a while and it "feels good", which is one of the main concerns that I had about this new resolution. So that's out of the way. Now I need to re-review the source code, and write some specs for it. Thank you for your patience (I know I am not the fastest guy doing these things)

buckle2000 commented:
...

kikito commented:
hi, sorry it is taking me so long :/

Last week I started working on this. The branch where I am doing stuff is here: https://github.com/kikito/bump.lua/tree/buckle2000-master

Some notes:

* I have renamed variables and used camelCase instead of under_score
* I have done some specs, but not finished them yet - so far they are identical to slide. I think I need some extra case to test the "special properties" of bypass.
* I am not super enthused about bypass calling slide. If you look at how the rest of the responses are implemented, you will see that all of them rely on recalculating a new goalX and goalY, and then returning `world:project`. I would really really like if bypass adhered to this (it's possible that we'll need to share some code between slide and bypass). I am having trouble because I still don't know exactly how bypass works (the double -corner thing is a bit disorienting. But I will figure it out)

That's why this isn't merged yet. Sorry it is taking me so long.

buckle2000 commented:
Thank you for your work!

"bypass" does not change the distance(scalar) of moving if not necessary. It moves object perpendicular to its original moving direction.

I tried to calculate goalX and goalY directly but that version of code is glitchy and unusable.

Some thoughts on special variables:
- `x` and `y`: offset displacement caused by this response, compare to when no response is used
- `bypassed items` list-like table, if the object actually bypass anything

kikito commented:
Risking sounding like a broken record, we have switched years and I still have not accepted this pr. You can see the refactors I did on [my branch](https://github.com/kikito/bump.lua/tree/buckle2000-master). My current problem is that I have detected a place where this resolution seems to give troublesome results. It is here:

![screen shot 2017-01-06 at 13 41 05](https://cloud.githubusercontent.com/assets/63131/21718165/d12258f0-d415-11e6-83b1-2293ccfd7859.png)

The problem is that when I am on that position and I move diagonally up-left or up-right, the green box enters some of the red boxes and "teleports" like this:

![screen shot 2017-01-06 at 13 42 46](https://cloud.githubusercontent.com/assets/63131/21718196/21383ada-d416-11e6-93d1-68db58ee341f.png)
![screen shot 2017-01-06 at 13 43 06](https://cloud.githubusercontent.com/assets/63131/21718197/213d42dc-d416-11e6-90ea-21f479ecab4a.png)

These "teleports" don't occur when using plain 'slide', so this is a bit of a mistery.

I need to do some careful debuggging of what's happening. But lately I have been struggling to find some free time where my mental capacities are strong enough (I'm in a bit of crunch mode at work, working during weekends, and at night I'm too tired to code). My deadline is next week, so I should be able to go back to testing this soon. Hopefully.

buckle2000 commented:
Thank you for all your work!

mtdowling commented:
I implemented a similar approach in my game that's using bump, and I ran into the same type of teleportation scenario that you are running into. For me, I worked around it by ensuring that the compensation for attempting to move around corners never exceeded the initial dx and dy. This resulted in eliminating teleportation, however, the trade-off is that this can cause a sort of "vibration" effect if you continually attempt to move in the same direction.

kikito commented:
@mtdowling hi, could you share the code you used? I have been trying this for a while unsuccessfully. It's a bit demoralizing.

mtdowling commented on bump.lua:
```diff
@@ -194,6 +194,11 @@ local function rect_detectCollision(x1,y1,w1,h1, x2,y2,w2,h2, goalX, goalY)
   }
 end
 
+-- in the shadow of at least one side of rect
```
What does "shadow" mean in this context? Can we use more precise wording here?

kikito replied:
If you stretch the edges of a rectangle to the infinite, that defines a "cross-like" region. 

So this:
```
                                                                          
   ******                                                                 
   *    *                                                                 
   *    *                                                                 
   *    *                                                                 
   ******                                                                 
             
```

Has the following "shadow":
```
   *    *                                                                 
   *    *                                                                 
   *    *                                                                 
 *******************                                                      
   *    *                                                                 
   *    *                                                                 
   *    *                                                                 
 *******************                                                      
   *    *                                                                 
   *    *                                                                 
   *    *                                                                 
   *    *                                                                 
```        

`shadow` means "the other rectangle is touching this rectangle's cross".

mtdowling commented:
I didn't implement my version using a filter, though I think a filter is more appropriate. I published a gist with what I'm doing in my game. I tried to extract out the part that's relevant to this issue... I don't know if it'll help you, but feel free to take a look at my hacky collision rounding code: https://gist.github.com/mtdowling/2bd0c31ed01d1c83dd1024f758a14460.

If you could add specs that fail for this PR, I could see if I can figure out how to fix them. Otherwise, I'm not really sure what an easy way of testing this out would be (short of making a simple little game).

kikito commented:
Thanks a lot, I will give this a look.

kikito commented:
This has become stale and if I ever come back to bump.lua I would probably start by making it much smaller - perhaps removing responders completely (which I understand is the most useful feature for a lot of people). What a conumdrum.

I am grateful for the time everyone put into this PR. Sometimes however it is better to let things go.

I won't feel insulted at all if anyone decides to fork and include this on their fork.
